aidistillery/graph_embeddings.py

```diff
@@ -0,0 +1,222 @@
+    """
+    Args
+    ====
+    probs: array of probabilities to sort
+    subset: [(idx, item)]
+
+    Ret
+    ===
+    Top-n of (item, prob) pairs.
+    """
+
+    zipped_probs = [(item, probs[idx]) for idx, item in subset]
+    zipped_probs.sort(key=itemgetter(1), reverse=True)
+    return zipped_probs[:n]
+
+def pagerank_main(args):
+    db = load_data(args.picklefile)
+    g, idx2doc, idx2tag, idx2aut = build_graph(db, directed=False)
+    print("%d nodes, %d edges." % (g.number_of_nodes(), g.number_of_edges()))
+
+    N = len(g)
+    DAMP = args.damp
+
+    print("N =", N)
+    print("DAMP =", DAMP)
+
+    def pagerank_builtin(g):
+        g.ndata['pv'] = g.ndata['pv'] / g.ndata['deg']
+        g.update_all(message_func=fn.copy_src(src='pv', out='m'),
+                     reduce_func=fn.sum(msg='m',out='m_sum'))
+        g.ndata['pv'] = (1 - DAMP) / N + DAMP * g.ndata['m_sum']
+
+    print("Running pagerank for %d steps" % args.steps)
+    # Initial values
+    g.ndata['pv'] = th.ones(N) / N
+    g.ndata['deg'] = g.out_degrees(g.nodes()).float()
+    for k in tqdm(range(args.steps)):
+        pagerank_builtin(g)
+
+    print("Sorting...")
+    # Only care for paper nodes, not tags or authors
+    pv = g.ndata['pv']
+    top_papers = top(pv, enumerate(idx2doc), args.top)
+    print("Top %d hot papers by PageRank:" % args.top)
+    for k, (key, score) in enumerate(top_papers):
+        suppl_data = db[key]
+        print("Top-{} (pv={:.2E}): '{}' ({}, {}, {} authors, Tags: {})".format(
+            k+1, score,
+            ' '.join(suppl_data['title'].split()),
+            suppl_data['link'],
+            suppl_data['published_parsed'].tm_year,
+            len(suppl_data['authors']),
+            ', '.join(t['term'] for t in suppl_data['tags'])))
+
+    print("Top %d hot authors" % args.top)
+    top_authors = top(pv, idx2aut.items(), args.top)
+    for k, (key, score) in enumerate(top_authors):
+        print("Top-{} (pv={:.2E}): {}".format(k+1, score, key))
+
+    print("Top %d hot tags" % args.top)
+    top_tags = top(pv, idx2tag.items(), args.top)
+    for k, (key, score) in enumerate(top_tags):
+        print("Top-{} (pv={:.2E}): {}".format(k+1, score, key))
+
+    print("kthxbye.")
+
+
+def pagerank_add_args(parser):
+    parser.add_argument('picklefile', help="Path to pickled input data")
+    parser.add_argument('-k',
+                        '--steps',
+                        metavar='K',
+                        help="Number of pagerank steps [100]",
+                        default=100,
+                        type=int)
+
+    parser.add_argument('-n',
+                        '--top',
+                        metavar='N',
+                        help="Number of top page-ranked items to return [10]",
+                        default=10,
+                        type=int)
+
+    parser.add_argument('-d', '--damp', help="Damping Factor",
+                        default=.85, type=float)
+
+def gcn_main(args):
```

Comment:
I assume these are not ready yet but you kept them to motivate you to do gcn? :D